pkg/fuzzer/fuzzer.go

```diff
@@ -68,67 +65,105 @@ func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
 	return f
 }
 
-type Config struct {
-	Debug          bool
-	Corpus         *corpus.Corpus
-	Logf           func(level int, msg string, args ...interface{})
-	Coverage       bool
-	FaultInjection bool
-	Comparisons    bool
-	Collide        bool
-	EnabledCalls   map[*prog.Syscall]bool
-	NoMutateCalls  map[int]bool
-	FetchRawCover  bool
-	NewInputFilter func(call string) bool
+type execQueues struct {
+	smashQueue           *queue.PlainQueue
+	triageQueue          *queue.PriorityQueue
+	candidateQueue       *queue.PlainQueue
+	triageCandidateQueue *queue.PriorityQueue
+	source               queue.Source
 }
 
-type Request struct {
-	Prog       *prog.Prog
-	NeedSignal SignalType
-	NeedCover  bool
-	NeedHints  bool
-	// If specified, the resulting signal for call SignalFilterCall
-	// will include subset of it even if it's not new.
-	SignalFilter     signal.Signal
-	SignalFilterCall int
-	// Fields that are only relevant within pkg/fuzzer.
-	flags   ProgTypes
-	stat    *stats.Val
-	resultC chan *Result
+func newExecQueues(fuzzer *Fuzzer) execQueues {
+	ret := execQueues{
+		triageCandidateQueue: queue.Priority(),
+		candidateQueue:       queue.PlainWithStat(fuzzer.StatCandidates),
+		triageQueue:          queue.Priority(),
+		smashQueue:           queue.Plain(),
+	}
+	// Sources are listed in the order, in which they will be polled.
+	ret.source = queue.Multiplex(
+		ret.triageCandidateQueue,
+		ret.candidateQueue,
+		ret.triageQueue,
+		// Alternate smash jobs with exec/fuzz in 33% of cases.
+		queue.Alternate(ret.smashQueue, fuzzer.rand(), 1.0/3.0),
+		queue.Callback(fuzzer.genFuzz),
+	)
+	return ret
 }
 
-type SignalType int
+type execOpt any
+type dontTriage struct{}
+type progFlags ProgTypes
 
-const (
-	NoSignal  SignalType = iota // we don't need any signal
-	NewSignal                   // we need the newly seen signal
-	AllSignal                   // we need all signal
-)
+func (fuzzer *Fuzzer) validateRequest(req *queue.Request) {
+	if req.NeedHints && (req.NeedCover || req.NeedSignal != queue.NoSignal) {
+		panic("Request.NeedHints is mutually exclusive with other fields")
+	}
+	if req.SignalFilter != nil && req.NeedSignal != queue.NewSignal {
+		panic("SignalFilter must be used with NewSignal")
+	}
+}
+
+func (fuzzer *Fuzzer) execute(executor queue.Executor, req *queue.Request, opts ...execOpt) *queue.Result {
+	fuzzer.validateRequest(req)
+	executor.Submit(req)
+	res := req.Wait(fuzzer.ctx)
+	fuzzer.processResult(req, res, opts...)
+	return res
+}
+
+func (fuzzer *Fuzzer) prepare(req *queue.Request, opts ...execOpt) {
+	fuzzer.validateRequest(req)
+	req.OnDone(func(req *queue.Request, res *queue.Result) bool {
+		fuzzer.processResult(req, res, opts...)
```

Comment:
Overall I like the new architecture. The last thing I am starring at is somewhat conflicting/overlapping in functionality Wait/OnDone features.
Formally, this use of OnDone is wrong, with all the generality we don't know if there is another OnDone callback already installed that will abort done handling later for some reason (e.g. some quality checker, which will say the result is low quality, let's discard and retry it). Formally it must use Wait for processing that must happen we decided the result won't be aborted.
The same issue holds for Deduplicator, it would be a good use for OnDone, except that Deduplicator wants a notification when the done is guaranteed to not be aborted.

But this "I want OnDone callback that can abort processing" vs "I want OnDone that cannot be aborted" is orthogonal to synchronous waiting vs async OnDone handling. Some caller, e.g. execute above do want to block and wait for completion.

Not sure what exactly we need to do here.
We could add another OnDone flavor for callbacks that are executed at the end of the chain and don't return bool (OnComplete or OnFinish). But we still need synchronous waiting. Potentially waiting can be a wrapper around OnComplete/OnFinish, but it looks ugly from API point of view.

Reply:
Another alternative I've been thinking about is a totally synchronous interface like

```
type Executor interface {
  Execute(*Request) *Result
}
```

And only at the last stage (in `rpc.go`) we'd introduce a queue and actually turn Request into Result. That would make all intermediate transformations much much easier to write and we don't need any `Wait()` and `OnDone()` methods.

But:
* It's very unclear how to order requests then, because we do want to prioritize them. Looks like we'd still have to introduce some intermediate queues + some goroutines that would explicitly take requests from those queues, in the order, and push them to the next stage. Too much Request copy/pasting then?
* There are `exec fuzz` and `exec gen` requests, which are kind of "last resort" activities, very different from the active entitites like jobs or machine checks. We'd need some exec fuzz/gen workers and we'd need to adjust their count to the speed with which we could execute them.
* For candidates, we'd also have to either create a goroutine for every candidate that'd just call `Execute(program)`, and there'd be tens of thousands of them, or again we'd have to spin up we don't know how many workers.

Reply:
Okay, `Request` ordering may be solved by assigning the priority value to each Request in a wrapper (like it's currently done in `queue.PriorityQueue()`) and then putting them all into one big priority queue somewhere close to where it's actually executed (e.g. in `rpc.go`). We can assign these priorities dynamically and build whatever orders we want to.

`exec fuzz` and `exec gen` should theoretically be solvable -- it's not a very difficult task to adjust the number of workers dynamically.

But I still don't see how we could do `skip every 3rd exec smash` there.

Reply:
Taking into account problems with candidates, fuzz/gen, smash, it does not look like a strict improvement. I don't like when abstraction impose unwanted runtime behavior, good abstraction they should support expressing things users want to express.

The current design looks very reasonable. The comments I had are really about some local inconveniences. So so far it looks like if we get maybe few local improvements to the current interfaces, we can get something that good from all points of view.

Reply:
As discussed offline, the current approach seems to make implementation easier.